middle_end/flambda/inlining/inlining_decision.ml

```diff
@@ -423,23 +423,63 @@ let might_inline dacc ~apply ~function_decl ~simplify_expr ~return_arity
 
 let make_decision_for_call_site dacc ~simplify_expr ~function_decl
       ~apply ~return_arity : Call_site_decision.t =
+  let function_decl_rec_info = I.rec_info function_decl in
+  let function_decl_rec_info =
+    match
+      Flambda_type.prove_rec_info (DA.typing_env dacc) function_decl_rec_info
+    with
+    | Proved rec_info -> rec_info
+    | Unknown -> Rec_info_expr.unknown
+    | Invalid -> Rec_info_expr.do_not_inline
+  in
   let inline = Apply.inline apply in
   match inline with
   | Never_inline -> Never_inline_attribute
   | Default_inline | Unroll _ | Always_inline | Hint_inline ->
-      if Inlining_state.is_depth_exceeded (Apply.inlining_state apply)
+    let unrolling_depth =
+      Simplify_rec_info_expr.known_remaining_unrolling_depth
+        dacc function_decl_rec_info
+    in
+    match unrolling_depth with
+    | Some 0 ->
+      Unrolling_depth_exceeded
+    | Some _ ->
+      might_inline dacc ~apply ~function_decl ~simplify_expr ~return_arity
```

Comment:
Could you leave a comment to explain that at this point if `unrolling_depth` is `Some _` it means that the user annotated this function with an unroll annotation earlier on, meaning that we are sure that this function should be unrolled?

Reply:
Sure, done.